cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrade.java

```diff
@@ -0,0 +1,195 @@
+    // skip default
+    keyParts.skipString();
+    String appId = keyParts.getString();
+    String programId = keyParts.getString();
+    String pId = keyParts.getString();
+
+    defaultStore.setStart(Id.Program.from(Constants.DEFAULT_NAMESPACE, appId, programId), pId, runRecord.getStartTs());
+  }
+
+  /**
+   * Handled the {@link AppMetadataStore#TYPE_RUN_RECORD_COMPLETED} meta data and writes it back with namespace
+   *
+   * @param row the {@link Row} containing the {@link AppMetadataStore#TYPE_RUN_RECORD_COMPLETED} metadata
+   */
+  private void runRecordCompletedHandler(Row row) {
+    RunRecord runRecord = GSON.fromJson(Bytes.toString(row.get(COLUMN)), RunRecord.class);
+    MDSKey.Splitter keyParts = new MDSKey(row.getRow()).split();
+
+    // skip runRecordStopped String
+    keyParts.skipString();
+    // skip default
+    keyParts.skipString();
+    String appId = keyParts.getString();
+    String programId = keyParts.getString();
+    long startTs = keyParts.getLong();
+    String pId = keyParts.getString();
+
+    writeTempRunRecordStart(appId, programId, pId, startTs);
+    defaultStore.setStop(Id.Program.from(Constants.DEFAULT_NAMESPACE, appId, programId), pId, runRecord.getStopTs(),
+                         ProgramController.State.getControllerStateByStatus(runRecord.getStatus()));
+  }
+
+  /**
+   * Writes the {@link AppMetadataStore#TYPE_RUN_RECORD_STARTED} entry in the app meta table so that
+   * {@link AppMetadataStore#TYPE_RUN_RECORD_COMPLETED} can be written which deleted the started record.
+   */
+  private void writeTempRunRecordStart(String appId, String programId, String pId, long startTs) {
+    defaultStore.setStart(Id.Program.from(Constants.DEFAULT_NAMESPACE, appId, programId), pId,
+                          Long.MAX_VALUE - startTs);
+  }
+
+  /**
+   * Writes the updated application metadata through the {@link DefaultStore}
+   *
+   * @param row the {@link Row} containing the application metadata
+   * @throws URISyntaxException if failed to create {@link URI} from the archive location in metadata
+   */
+  private void applicationMetadataHandler(Row row) throws URISyntaxException {
+    ApplicationMeta appMeta = GSON.fromJson(Bytes.toString(row.get(COLUMN)), ApplicationMeta.class);
+    defaultStore.addApplication(Id.Application.from(Constants.DEFAULT_NAMESPACE, appMeta.getId()), appMeta.getSpec(),
+                                locationFactory.create(
+                                  new URI(updateAppArchiveLocation(appMeta.getId(), appMeta.getArchiveLocation()))));
+  }
+
+  /**
+   * Creates the new archive location for application with namespace
+   *
+   * @param appId           : the application id
+   * @param archiveLocation : the application archive location
+   * @return the new archive location with namespace
+   */
+  private String updateAppArchiveLocation(String appId, String archiveLocation) {
+    return archiveLocation.replace("programs/archive" + FORWARD_SLASH + DEVELOPER_STRING,
```

Comment:
Can we use ``String.format`` to construct the two Strings/CharSequences here separately and then call ``replace``? Very hard to read this way.